### ISSUE TITLE
julia: before_script suppresses default script?

### ISSUE THREAD

**Issue by wavexx** [travis-ci/travis-ci#3988]
I've tried to use "before_script" to setup passwordless ssh access to localhost, as I normally did in Python. I'm not redefining the default "script" target, since I'm fine with the default.

See:

https://github.com/wavexx/Polyglot.jl/blob/1e324d3b30032a4f371d59f96054d1579e9e6167/.travis.yml

However, when I define before_script, it looks like the default Julia "script" is not run at all.
See the resulting build log:

https://travis-ci.org/wavexx/Polyglot.jl/jobs/64701128

I also tried to add that to "before_install" and "install".
It seems that as soon as one of the targets is defined I lose the convenience of the pre-defined rules.

**Comment by BanzaiMan** [comment]
Paging @tkelman @ninjin @staticfloat

**Comment by tkelman** [comment]
Oh! This was a consequence of you renaming the package and repo, at the commit as of that build [the `src` directory](https://github.com/wavexx/Polyglot.jl/tree/4b9b2392b72e9a2dad34b2c8b70305f676eb1fec/src) contained a `Bond.jl` but not a `Polyglot.jl`, so the Travis build scripts didn't think your repository was a standard Julia package and didn't know what to do with it.

We should maybe add a few informational printouts, "Package name guessed from repository name to be $JL_PKG", and a message in this condition when `src/$JL_PKG.jl` does not exist.

**Comment by tkelman** [comment]
I don't really know ruby, but can `sh.if "-f src/${JL_PKG}.jl" do` have an else clause added to it? We could echo a message in that condition, "`src/${JL_PKG}.jl` not found, this repository is not a valid Julia package so the default test script is empty."

**Comment by wavexx** [comment]
On 05/30/2015 07:51 PM, Tony Kelman wrote:
> Oh! This was a consequence of you renaming the package and repo, at
> the commit as of that build [the `src`
> directory](https://github.com/wavexx/Polyglot.jl/tree/4b9b2392b72e9a2dad34b2c8b70305f676eb1fec/src)
> contained a `Bond.jl` but not a `Polyglot.jl`, so the Travis build
> scripts didn't think your repository was a standard Julia package and
> didn't know what to do with it.

In subsequent commits I did rename Bond.jl to Polyglot.jl (I
force-pushed the tests so they don't show up in git). It didn't make a
difference for travis though.

I also ended up discovering that $JL_PKG was still undefined, so in my
latest travis.yml I defined it manually along with the script.

Is it possible that travis is still using some old metadata after the
rename?

coveralls for instance required me to remove the old name and enable the
new one.

> We should maybe add a few informational printouts, "Package name
> guessed from repository name to be $JL_PKG", and a message in this
> condition when `src/$JL_PKG.jl` does not exist.

Printing the discovered package name would be very useful for debugging
indeed.

**Comment by BanzaiMan** [comment]
@tkelman `if …; else…; fi` construct is easy. For example, https://github.com/travis-ci/travis-build/blob/c1748ffd239eac0702d546b3f988dc88963ee0fb/lib/travis/build/script/node_js.rb#L35-42 shows how to create an equivalent of:

```bash
if [ -f package.json ]; then
  npm test
else
  make test
fi
```

**Comment by tkelman** [comment]
Thanks @BanzaiMan, see https://github.com/travis-ci/travis-build/pull/459

**Comment by wavexx** [comment]
Ok, I tried again and indeed, it works as expected.
However, is "JL_PKG" available in "before_script"? I tried an "echo", but it turned out empty.

**Comment by tkelman** [comment]
Don't think so, since `set_jl_pkg` runs during `script` - https://github.com/travis-ci/travis-build/blob/18bd04e965b9bfaa49cd6bdcd8dcb1513b8d2fcd/lib/travis/build/script/julia.rb#L61

**Comment by wavexx** [comment]
On 06/01/2015 12:14 PM, Tony Kelman wrote:
> Don't think so, since `set_jl_pkg` runs during `script` -
> https://github.com/travis-ci/travis-build/blob/18bd04e965b9bfaa49cd6bdcd8dcb1513b8d2fcd/lib/travis/build/script/julia.rb#L61

Wouldn't it make sense to define it earlier? Just after the cloning of
the package, and probably before exporting variables (so that we can
override the local environment if needed).

I would also document JL_PKG more publicly. I'm using it in
after_success to publish coverage as well:

after_success:
   - if [ $TRAVIS_JULIA_VERSION = "nightly" ]; then julia -e
"cd(Pkg.dir(\"${JL_PKG}\")); Pkg.add(\"Coverage\"); using Coverage;
Coveralls.submit(Coveralls.process_folder())"; fi

This seems a pretty common rule grepping through my ~/.julia folder,
except that everybody manually adds their own pkg name,

Maybe we can can configure the rule somehow, with a new block?

submit:
  coveralls: nightly/stable/yes

(where yes means both) and then provide a default after_success as well.

**Comment by tkelman** [comment]
I agree we probably want to do some automation for coveralls since a bunch of packages are doing it manually right now, but that's a separate issue. See #3092 and https://github.com/JuliaCI/travis-build/issues/1. Implementation suggestions welcome, especially if you know ruby better than I do.

Right now `JL_PKG` needs to be determined after Julia gets installed, since we're using the same regex as `Pkg` in base Julia. At least that's the idea. We could maybe switch to running it through perl or ruby or some other pre-installed language to do it much earlier, as long as we're sure we'd get the same result.

**Comment by wavexx** [comment]
The expression seems pretty trivial. No need to use Julia for that:

``git remote -v | head -n 1 | cut -f 2 | cut -f 1 -d ' ' | sed -rne 's/(.*[\/\\])?(\w+?)(\.jl)?(\.git)?$/\2/p'``

**Comment by wavexx** [comment]
As an additional note, no, I barely ever used ruby at all, so I'd rather let somebody else chime in.

**Comment by ninjin** [comment]
Agreed that the expression is simple.  The reason for hooking into Julia is that the regular expression is actually the official definition of what is a valid package name.  About Ruby, I should mention that the only Ruby code that I have ever written is the Julia patch for travis-ci.

**Comment by tkelman** [comment]
Oh. Hah. Well Julia and Ruby aren't all that different syntax-wise, but it helps to be aware of what the differences are.

**Comment by wavexx** [comment]
I imagined, but since we pulled that over from the source, I'm not so sure about requiring julia just to run it. It's basically a basename of the remote with .jl(.git) stripped. If the regex fails, it expands to nothing.

I've run it though my julia folder to test it:

    for dir in `ls -d .julia/*/*`; do test -d $dir && (cd $dir; git remote -v | head -n 1 | cut -f 2 | cut -f 1 -d ' ' | sed -rne 's/(.*[\/\\])?(\w+?)(\.jl)?(\.git)?$/\2/p'; ); done

and it seems to work as expected.

**Comment by tkelman** [comment]
Is https://github.com/travis-ci/travis-build/pull/459 sufficient to close this? I tested it out on my own package and both the new echo's appear to be working correctly. (Though `:yellow` and `:green` look about the same on my display...)

We could move around `set_jl_pkg` and more thoroughly document `JL_PKG` if we want, but I don't see that as especially high priority?

**Comment by wavexx** [comment]
On 06/02/2015 11:43 PM, Tony Kelman wrote:
> Is https://github.com/travis-ci/travis-build/pull/459 sufficient to
> close this? I tested it out on my own package and both the new echo's
> appear to be working correctly. (Though `:yellow` and `:green` look
> about the same on my display...)

Good for me.

> We could move around `set_jl_pkg` and more thoroughly document
> `JL_PKG` if we want, but I don't see that as especially high
> priority?

Agreed, it's not high priority.
Though I would maybe create an issue for it.
Having JL_PKG in before_script, at least, would make sense to me.
